src/security/CertGadgets.h

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright (C) 1996-2021 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#ifndef SQUID_SRC_SECURITY_CERTGADGETS_H
+#define SQUID_SRC_SECURITY_CERTGADGETS_H
+
+#include "security/forward.h"
+
+namespace Security
+{
+
+/// \return the SubjectName field of the given certificate
+SBuf CertSubjectName(const CertPointer &);
+
+/// \return the Issuer field of the given certificate
+SBuf CertIssuerName(const CertPointer &);
```

Comment:
Done in 611b46a.

Reply:
I have take a look at using the Certificate type as parameters here. It results in every caller having to de-references a Pointer to pass the parameter, then the function itself immediately having to generate a pointer to the object to pass on to the library APIs.

While it may *look* like a good API, it is simpler and more efficient code just to pass the Pointer in.

Reply:
I counted six statements in your rebuttal. AFAICT, at least three of them are factual errors or misrepresentations. I will need more time to analyze this situation further and respond constructively.

Reply:
> The library API provides us with raw-pointer values.

Yes. They have no better choice because C does not have references.

> * As a result of that,  all the callers are handling CertPointer.

Callers that store certificates for long-term use do. Callers that manipulate _existing_ certificates (given to them) should be handling certificates, not never-nil pointers to those certificates.

> * As a result of that, in order to make these function/method API take a Certificate& instead of CertPointer& the callers all would have to de-reference the CertPointer 

Yes, those who store pointers will have to dereference them to examine stored certificates, but that is not a problem.

> like so:...

The actual code will usually look better than the suggested sketch, but the specific dereferencing code does not really matter here. What matters is whether the caller may have a nil pointer, whether the function has a meaningful outcome when given nil pointers, whether that outcome should be reported differently from other problematic outcomes, etc.


> The library APIs all take raw-pointer to the certificate type. So the CertIssuerName(Certificate&) and CertSubjectName(Certificate&) functions have to take a pointer to the Certificate& they receive when calling the library.

Yes, but that is a very minor overhead compared to the advantages offered by the correct Squid API.

> Which makes our code bunch of unnecessary sequences of

Similar sequences are often _necessary_ to handle nil pointers correctly anyway. They result in very little (sometimes zero) runtime overhead, so it is the price worth paying for the code correctness/safety.


> IMO, there is no problem using APIs with (const CertPointer &) when the objects we are passing around are otherwise raw-pointers.

The primary problem is the meaning and handling of nil pointers, especially where the library functions _crash_ on nil pointers (e.g., X509_get_issuer_name() and gnutls_x509_crt_get_issuer()) and where the functions test some condition that may be negated (e.g., "A expires before B"). When Squid code can simply feed the pointer to the library, we are very likely to succumb to the temptation to do so. In some cases, such feeding will result in misleading return values, crashes, assertions, or UB.

There are also secondary problems like excessive storage/leaks of allocated certificates in code where no storage should be happening at all -- it is trivial to store a certificate using a pointer while storing it using a reference requires more work that is more likely to be noticed by a reviewer. The current PR code (as of 516ed97) suffers from this excessive storage problem, for example.

Whenever we violate a basic development/engineering principle such as "do not use pointers for must-exist objects", we must have a very good justification/reason for doing so. Avoiding a `&` character at the C++/C boundary is not a very good justification/reason!

FWIW, I see two options for addressing this change request:

1. You try to adjust PR APIs in question to take references and sync the affected code accordingly.
2. I do the above.

Regardless of who does this legwork, we review the resulting code to confirm that the revised approach is overall better.

This first option is probably faster overall, provided I have convinced you of the merits of using these APIs so that you can faithfully implement the necessary changes.

Reply:
I converted CertSubjectName() and CertIssuerName() in branch commits 1b90b8e and e6234a4. Please see e6234a4 for a potentially interesting/important GnuTLS caveat. The branch code will need some polishing, but that should be done later IMO.

So far, I have not noticed any cascading changes or other problems you were [worried](https://github.com/squid-cache/squid/pull/831#discussion_r736098457) about. One branch bug fixed as a side effect (commit 1b90b8e).

Next, I will find the time to convert CertIsIssuedBy() in a similar fashion.

Reply:
> Next, I will find the time to convert CertIsIssuedBy() in a similar fashion.

Done in commit e93e4e4. I also polished PR changes a little in commit 3109379. The other two recent commits just fix the side effects of those changes, although commit 350dc38 is rather interesting (and was difficult for me to figure out).

I am resolving this change request as addressed.